napari/utils/actions/_registries.py

```diff
@@ -0,0 +1,326 @@
+from __future__ import annotations
+
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import TYPE_CHECKING, Callable, Optional
+
+from psygnal import Signal
+
+from ...utils.translations import trans
+from ._menus import MenuGroup, MenuId
+from ._types import (
+    MenuItem,
+    SubmenuItem,
+    _RegisteredCommand,
+    _RegisteredKeyBinding,
+)
+from ._util import MockFuture
+
+if TYPE_CHECKING:
+    from typing import Dict, Iterator, List, Sequence, Set, Tuple, Union
+
+    from ._types import (
+        CommandHandler,
+        CommandId,
+        CommandRule,
+        KeybindingRule,
+        TranslationOrStr,
+    )
+
+    DisposeCallable = Callable[[], None]
+    CommandDecorator = Callable[[CommandHandler], CommandHandler]
+    MenuOrSubmenu = Union[MenuItem, SubmenuItem]
+
+
+class CommandsRegistry:
+
+    registered = Signal(str)
+    __instance: Optional[CommandsRegistry] = None
+
+    def __init__(self) -> None:
+        self._commands: Dict[CommandId, List[_RegisteredCommand]] = {}
+
+    @classmethod
+    def instance(cls) -> CommandsRegistry:
+        if cls.__instance is None:
+            cls.__instance = cls()
+        return cls.__instance
+
+    def register_command(
+        self,
+        id: CommandId,
+        callback: CommandHandler,
+        title: TranslationOrStr = '',
+    ) -> DisposeCallable:
+        """Register a callable as the handler for command `id`.
+
+        Parameters
+        ----------
+        id : CommandId
+            Command identifier
+        callback : Callable
+            Callable to be called when the command is executed
+        title : TranslationOrStr
+            Optional title for the command.
+
+        Returns
+        -------
+        DisposeCallable
+            A function that can be called to unregister the command.
+        """
+        commands = self._commands.setdefault(id, [])
+
+        cmd = _RegisteredCommand(id, callback, title)
+        commands.insert(0, cmd)
+
+        def _dispose():
+            commands.remove(cmd)
+            if not commands:
+                del self._commands[id]
+
+        self.registered.emit(id)
+        return _dispose
+
+    def __iter__(self) -> Iterator[Tuple[CommandId, List[_RegisteredCommand]]]:
+        yield from self._commands.items()
+
+    def __contains__(self, id: str) -> bool:
+        return id in self._commands
+
+    def __repr__(self) -> str:
+        name = self.__class__.__name__
+        return f"<{name} at {hex(id(self))} ({len(self._commands)} commands)>"
+
+    def __getitem__(self, id: CommandId) -> List[_RegisteredCommand]:
+        """Retrieve commands registered under a given ID"""
+        return self._commands[id]
```

Comment:
```suggestion
        return self._commands[id][:]
```
copy for safety? Otherwise, someone could modify the inner state of register?